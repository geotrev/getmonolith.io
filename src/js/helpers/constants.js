export const KeyCodes = {
  SHIFT: 16,
  TAB: 9,
  ARROW_UP: 38,
  ARROW_DOWN: 40,
}

export const Selectors = {
  FOCUSABLE_TAGS: ["a", "button", "input", "object", "select", "textarea", "[tabindex]"],
  KEYBOARD_CLASS: "using-keyboard",
  NOT_VISUALLY_HIDDEN_CLASS: ":not(.is-visually-hidden)",
  TABINDEX: "tabindex",
}

export const Events = {
  KEYDOWN: "keydown",
  CLICK: "click",
  BLUR: "blur",
}

export const Messages = {
  NO_SELECTOR_STRING_OR_CHILDREN_ERROR:
    "createFocusTrap must be given one or both of: first parameter (as selector string)" +
    " and/or options.children (array of elements).",
  OPTION_MATCHERS_DATA_TYPE_ERROR:
    "Invalid data type given to options.matchers for createFocusTrap. Expected: Array.",
  INCORRECT_MATCHER_TYPE_ERROR: type =>
    `Invalid matcher given to options.matchers for createFocusTrap. Expected: String. Recieved: ${type}.`,
  NO_MATCHER_LENGTH_ERROR:
    "Invalid value given to options.matchers for createFocusTrap; value must be an array with at least one selector string",
}
