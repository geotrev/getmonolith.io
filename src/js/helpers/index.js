export * from "./component-engine"
export { createFocusRing } from "./create-focus-ring"
export { createFocusTrap } from "./create-focus-trap"
export * from "./dom"
export { focusOnce } from "./focus-once"
export { isBrowserEnv } from "./is-browser-env"
export { isiOSMobile } from "./is-ios-mobile"
export { log } from "./log"
export { throttle } from "./throttle"
