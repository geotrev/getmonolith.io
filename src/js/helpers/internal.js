/**
 * Private method and property symbols used in the base class, mixins, and components.
 * These are used to prevent external scripts from accessing the internals of Undernet's
 * core logic.
 */
export const internal = {
  // Component
  start: Symbol("start"),
  stop: Symbol("stop"),
  validate: {
    method: Symbol("validate"),
    property: Symbol("_validate"),
  },
  teardown: {
    method: Symbol("teardown"),
    property: Symbol("_teardown"),
  },
  components: {
    method: Symbol("components"),
    property: Symbol("_components"),
  },
  activeComponent: {
    method: Symbol("activeComponent"),
    property: Symbol("_activeComponent"),
  },
  cancelActiveComponent: {
    method: Symbol("cancelActiveComponent"),
    property: Symbol("_cancelActiveComponent"),
  },
  // Mixins
  isVisible: {
    method: Symbol("isVisible"),
    property: Symbol("_isVisible"),
  },
}
