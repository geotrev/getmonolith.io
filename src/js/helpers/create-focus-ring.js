import { Events, Selectors } from "./constants"
import { isBrowserEnv } from "./is-browser-env"

/**
 * Factory function that creates focus ring helpers.
 *
 * ```js
 * const focusRing = createFocusRing()
 * focusRing.start()
 * focusRing.stop()
 * ```
 *
 * @returns {{ start: Function, stop: Function }}
 */
export const createFocusRing = () => {
  if (!isBrowserEnv) return

  let listeningForKeydown

  const listenForKeyboard = () => {
    document.body.classList.add(Selectors.KEYBOARD_CLASS)
    document.removeEventListener(Events.KEYDOWN, listenForKeyboard)
    // eslint-disable-next-line no-use-before-define
    document.addEventListener(Events.CLICK, listenForClick)
    listeningForKeydown = false
  }

  const listenForClick = () => {
    document.body.classList.remove(Selectors.KEYBOARD_CLASS)
    document.removeEventListener(Events.CLICK, listenForClick)
    document.addEventListener(Events.KEYDOWN, listenForKeyboard)
    listeningForKeydown = true
  }

  return {
    start() {
      document.addEventListener(Events.KEYDOWN, listenForKeyboard)
    },
    stop() {
      if (listeningForKeydown) {
        document.removeEventListener(Events.KEYDOWN, listenForKeyboard)
      } else {
        document.body.classList.remove(Selectors.KEYBOARD_CLASS)
        document.removeEventListener(Events.CLICK, listenForClick)
      }
    },
  }
}
