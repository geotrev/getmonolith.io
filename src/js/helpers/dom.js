import { Selectors } from "./constants"

/**
 * Run a `querySelectorAll` but convert the result to an array literal for IE11 compat.
 * @param {string} selector
 * @param {HTMLElement} parent
 */
export const queryAll = (selector, parent = document) => {
  return Array.apply(null, parent.querySelectorAll(selector))
}

/**
 * Search for elements matching a given selector.
 *
 * ```js
 * const elements = getFocusableElements(".wrapper")
 * elements.forEach(element => element.classList.add("focusable"))
 * ```
 *
 * Use a custom pattern to select only specific elements:
 *
 * ```js
 * const elements = getFocusableElements(".wrapper", [".my-button"])
 * ```
 *
 * @param {String} selectorString - The selector string of the container element.
 * @param {String[]} matchers - Optional matchers override. Defaults to common focusable selectors.
 * @returns {Element[]} Static array of HTML elements
 */
export const getFocusableElements = (selectorString, matchers = Selectors.FOCUSABLE_TAGS) => {
  const focusables = matchers
    .map(selector => `${selectorString} ${selector}${Selectors.NOT_VISUALLY_HIDDEN_CLASS}`)
    .join(", ")

  return queryAll(focusables)
}
