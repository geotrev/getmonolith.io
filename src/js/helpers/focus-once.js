import { Events, Selectors } from "./constants"

/**
 * Focus a single element one time and teardown when unfocused.
 * @param {HTMLElement} element
 */
export const focusOnce = element => {
  const handleBlur = ({ target }) => {
    target.removeAttribute(Selectors.TABINDEX)
    target.removeEventListener(Events.BLUR, handleBlur)
  }

  element.setAttribute(Selectors.TABINDEX, "-1")
  element.focus()
  element.addEventListener(Events.BLUR, handleBlur)
}
