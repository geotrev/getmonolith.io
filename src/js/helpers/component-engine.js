import { queryAll } from "./dom"
import { isBrowserEnv } from "./is-browser-env"

const isString = value => typeof value === "string"
const isFunction = value => typeof value === "function"

export const ComponentEngine = {
  /**
   * Initialize an Undernet component globally or by id.
   * @param {Object} metadata
   * @property {string} id
   * @property {string} attribute
   * @property {Object} thisArg
   */
  start: (metadata = {}) => {
    if (!isBrowserEnv) return

    const { id, attribute, thisArg } = metadata

    if (id && isString(id)) {
      const instance = document.querySelector(`[${attribute}='${id}']`)
      if (!instance) return

      const validComponent = [instance].filter(thisArg._validate)[0]
      if (!validComponent) return

      thisArg._components.push(validComponent)
    } else if (!id && !thisArg._components.length) {
      const instances = queryAll(`[${attribute}]`)
      if (!instances.length) return

      const validComponents = instances.filter(thisArg._validate)
      thisArg._components = thisArg._components.concat(validComponents)
    } else {
      // attempted to .start() when .stop() wasn't run,
      // OR tried to instantiate a component that's already active.
    }
  },
  /**
   * Teardown an Undernet component globally or by id.
   * @param {Object} metadata
   * @property {string} id
   * @property {string} attribute
   * @property {Object} thisArg
   * @property {string} activeNodeKey
   * @property {*=} cancelActiveFn
   */
  stop: (metadata = {}) => {
    if (!isBrowserEnv) return

    const { id, attribute, thisArg, activeNodeKey, cancelActiveFn } = metadata

    if (id && isString(id)) {
      let targetIndex
      const instance = thisArg._components.filter((activeInstance, index) => {
        if (activeInstance.getAttribute(attribute) !== id) return false
        targetIndex = index
        return true
      })[0]

      if (!instance) return

      if (
        thisArg[activeNodeKey] &&
        instance === thisArg[activeNodeKey] &&
        isFunction(cancelActiveFn)
      )
        thisArg[cancelActiveFn]()

      thisArg._teardown(instance)
      thisArg._components.splice(targetIndex, 1)
    } else if (!id && thisArg._components.length) {
      if (thisArg[activeNodeKey] && isFunction(cancelActiveFn)) thisArg[cancelActiveFn]()

      thisArg._components.forEach(thisArg._teardown)
      thisArg._components = []
    }
  },
}
