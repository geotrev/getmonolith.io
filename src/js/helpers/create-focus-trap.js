import { isBrowserEnv } from "./is-browser-env"
import { Selectors, Messages, KeyCodes, Events } from "./constants"
import { log } from "./log"
import { getFocusableElements } from "./dom"
/**
 * Factory function that creates focus trap helpers.
 *
 * ```js
 * const focusTrap = createFocusTrap("#element-id")
 * focusTrap.start()
 * focusTrap.stop()
 * ```
 *
 * Pass an object in the second param to use other helpers for arrow trapping, custom children, and custom matchers.
 *
 * ```js
 * const focusTrap = createFocusTrap("#element-id", {
 *    useArrows: true,
 *    children: [],
 *    matchers: ['a', 'button', '.my-cool-element']
 * })
 * focusTrap.start()
 * ```
 *
 * @param {String} selectorString
 * @param {{ useArrows, children: (NodeList|Element[]), matchers: String[] }} options - useArrows is coerced to true/false.
 * @returns {{ start: Function, stop: Function }}
 */
export const createFocusTrap = (selectorString, options = {}) => {
  if (!isBrowserEnv) return
  let focusTrapActive = false
  const { useArrows, children, matchers = Selectors.FOCUSABLE_TAGS } = options

  if (!selectorString && !children.length) {
    log(Messages.NO_SELECTOR_STRING_OR_CHILDREN_ERROR)
    return
  }

  if (!Array.isArray(matchers)) {
    log(Messages.OPTION_MATCHERS_DATA_TYPE_ERROR)
    return
  } else if (matchers.length) {
    let hasBadMatcher = false

    matchers.forEach(matcher => {
      const type = typeof matcher
      if (type !== "string") {
        log(Messages.INCORRECT_MATCHER_TYPE_ERROR(type))
        hasBadMatcher = true
      }
    })

    if (hasBadMatcher) return
  } else if (!matchers.length) {
    log(Messages.NO_MATCHER_LENGTH_ERROR)
    return
  }

  const focusableChildren =
    children && children.length ? children : getFocusableElements(selectorString, matchers)
  const focusableFirstChild = focusableChildren[0]
  const focusableLastChild = focusableChildren[focusableChildren.length - 1]

  const handleFocusTrapWithTab = event => {
    const containerElement = document.querySelector(selectorString)
    const containerActive = document.activeElement === containerElement
    const firstActive = document.activeElement === focusableFirstChild
    const lastActive = document.activeElement === focusableLastChild

    if (!containerActive && !firstActive && !lastActive) return

    const tabKey = event.which === KeyCodes.TAB
    const shiftKey = event.which === KeyCodes.SHIFT || event.shiftKey
    const hasShift = shiftKey && tabKey
    const noShift = !shiftKey && tabKey

    if (hasShift && (firstActive || containerActive)) {
      event.preventDefault()
      focusableLastChild.focus()
    } else if (noShift && lastActive) {
      event.preventDefault()
      focusableFirstChild.focus()
    }
  }

  const focusNextChild = () => {
    for (let i = 0; i < focusableChildren.length; i++) {
      if (focusableChildren[i] === document.activeElement) {
        focusableChildren[i + 1].focus()
        break
      }
    }
  }

  const focusLastChild = () => {
    for (let i = 0; i < focusableChildren.length; i++) {
      if (focusableChildren[i] === document.activeElement) {
        focusableChildren[i - 1].focus()
        break
      }
    }
  }

  const handleFocusTrapWithArrows = event => {
    const firstActive = document.activeElement === focusableFirstChild
    const lastActive = document.activeElement === focusableLastChild
    const arrowUp = event.which === KeyCodes.ARROW_UP
    const arrowDown = event.which === KeyCodes.ARROW_DOWN

    if (arrowUp || arrowDown) {
      event.preventDefault()

      if (firstActive && arrowUp) {
        focusableLastChild.focus()
      } else if (lastActive && arrowDown) {
        focusableFirstChild.focus()
      } else if (arrowDown) {
        focusNextChild()
      } else if (arrowUp) {
        focusLastChild()
      }
    }
  }

  return {
    start() {
      // Already active, exit
      if (focusTrapActive) return

      // Now active
      focusTrapActive = true

      // Handle focus listeners
      if (useArrows) {
        document.addEventListener(Events.KEYDOWN, handleFocusTrapWithArrows)
      } else {
        document.addEventListener(Events.KEYDOWN, handleFocusTrapWithTab)
      }
    },
    stop() {
      // Nothing to stop, exit
      if (!focusTrapActive) return

      // Now inactive
      focusTrapActive = false

      // Remove focus listeners
      if (useArrows) {
        document.removeEventListener(Events.KEYDOWN, handleFocusTrapWithArrows)
      } else {
        document.removeEventListener(Events.KEYDOWN, handleFocusTrapWithTab)
      }
    },
  }
}
