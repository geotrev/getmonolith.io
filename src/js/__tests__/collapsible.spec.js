import Undernet from "../"
import { find, renderDOM } from "./helpers"

const dom = `
  <div class="collapsible" data-collapsible="collapsible-id">
    <h5>
      <button id="collapsible-trigger-id" class="collapsible-trigger" data-target="collapsible-id">
        Collapsible Trigger
      </button>
    </h5>
    <div class="collapsible-content" id="collapsible-id">
      <p class="has-p">
        Consectetur eiusmod laboris in non id tempor exercitation ipsum cupidatat magna
        ipsum ut voluptate. <a href="#">E pluribus unum.</a>
      </p>
    </div>
  </div>
`

describe("Collapsible", () => {
  let wrapper

  beforeEach(() => {
    wrapper = renderDOM(dom)
  })

  afterEach(() => {
    Undernet.Collapsibles.stop()
  })

  describe("API start", () => {
    it("sets attributes", () => {
      // When
      Undernet.Collapsibles.start()
      // Then
      expect(wrapper()).toMatchSnapshot()
    })

    it("does not open collapsible by default if [data-visible='false'] is set", () => {
      find("[data-collapsible]").setAttribute("data-visible", "false")
      // When
      Undernet.Collapsibles.start()
      // Then
      expect(wrapper()).toMatchSnapshot()
    })
  })

  describe("API stop + #handleClick", () => {
    it("does not toggle collapsible on click", () => {
      const trigger = find("button")
      // When
      Undernet.Collapsibles.start()
      Undernet.Collapsibles.stop()
      trigger.click()
      // Then
      expect(wrapper()).toMatchSnapshot()
    })
  })

  describe("#handleClick", () => {
    it("toggles collapsible on click", () => {
      const trigger = find("button")
      // When
      Undernet.Collapsibles.start()
      trigger.click()
      // Then
      expect(wrapper()).toMatchSnapshot()
    })
  })
})

const errorDom = (collapsibleId, collapsibleTriggerId, triggerId, contentId) => `
  <div class="collapsible" data-collapsible="${collapsibleId}">
    <h5>
      <button id="${triggerId}" class="collapsible-trigger" data-target="${collapsibleTriggerId}">
        Collapsible Trigger
      </button>
    </h5>
    <div class="collapsible-content" id="${contentId}">
      <p class="has-m">
        Consectetur eiusmod laboris in non id tempor exercitation ipsum cupidatat magna
        ipsum ut voluptate. <a href="#">E pluribus unum.</a>
      </p>
    </div>
  </div>
`

describe("Collapsible Warnings", () => {
  const collapsibleId = "collapsible-id"
  const unMatchedCollapsibleId = "collapsible-id-no-match"
  const triggerId = "collapsible-trigger-id"

  beforeAll(() => {
    console.warning = jest.fn()
  })

  afterEach(() => {
    Undernet.Collapsibles.stop()
  })

  it("throws error if [data-collapsible] attribute is missing its value", () => {
    // Given
    renderDOM(errorDom("", unMatchedCollapsibleId, triggerId, collapsibleId))
    // When
    Undernet.Collapsibles.start()
    // Then
    expect(console.warning).toBeCalledWith(
      "Could not initialize collapsible; you must include a value for the 'data-collapsible' attribute."
    )
  })

  it("throws error if [data-trigger] attribute can't be found", () => {
    // Given
    renderDOM(errorDom(collapsibleId, unMatchedCollapsibleId, triggerId, collapsibleId))
    // When
    Undernet.Collapsibles.start()
    // Then
    expect(console.warning).toBeCalledWith(
      `Could not find collapsible trigger with [data-target='${collapsibleId}']; you can't have a collapsible without a trigger.`
    )
  })

  it("throws error if trigger id can't be found", () => {
    // Given
    renderDOM(errorDom(collapsibleId, collapsibleId, "", collapsibleId))
    // When
    Undernet.Collapsibles.start()
    // Then
    expect(console.warning).toBeCalledWith(
      `Could not find id on collapsible trigger with [data-target='${collapsibleId}'].`
    )
  })

  it("throws error if collapsible content can't be found", () => {
    // Given
    renderDOM(errorDom(collapsibleId, collapsibleId, triggerId, unMatchedCollapsibleId))
    // When
    Undernet.Collapsibles.start()
    // Then
    expect(console.warning).toBeCalledWith(
      `Could not find collapsible content with id '${collapsibleId}'; you can't have a collapsible without content.`
    )
  })
})
