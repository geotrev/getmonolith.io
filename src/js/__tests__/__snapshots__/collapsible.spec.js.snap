// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapsible #handleClick toggles collapsible on click 1`] = `
<body>
  
  
  <div
    class="collapsible"
    data-collapsible="collapsible-id"
    data-visible="false"
  >
    
    
    <h5>
      
      
      <button
        aria-controls="collapsible-id"
        aria-expanded="false"
        class="collapsible-trigger"
        data-target="collapsible-id"
        id="collapsible-trigger-id"
      >
        
        Collapsible Trigger
      
      </button>
      
    
    </h5>
    
    
    <div
      aria-hidden="true"
      aria-labelledby="collapsible-trigger-id"
      class="collapsible-content"
      id="collapsible-id"
      style=""
    >
      
      
      <p
        class="has-p"
      >
        
        Consectetur eiusmod laboris in non id tempor exercitation ipsum cupidatat magna
        ipsum ut voluptate. 
        <a
          href="#"
          tabindex="-1"
        >
          E pluribus unum.
        </a>
        
      
      </p>
      
    
    </div>
    
  
  </div>
  

</body>
`;

exports[`Collapsible API start does not open collapsible by default if [data-visible='false'] is set 1`] = `
<body>
  
  
  <div
    class="collapsible"
    data-collapsible="collapsible-id"
    data-visible="false"
  >
    
    
    <h5>
      
      
      <button
        aria-controls="collapsible-id"
        aria-expanded="false"
        class="collapsible-trigger"
        data-target="collapsible-id"
        id="collapsible-trigger-id"
      >
        
        Collapsible Trigger
      
      </button>
      
    
    </h5>
    
    
    <div
      aria-hidden="true"
      aria-labelledby="collapsible-trigger-id"
      class="collapsible-content"
      id="collapsible-id"
    >
      
      
      <p
        class="has-p"
      >
        
        Consectetur eiusmod laboris in non id tempor exercitation ipsum cupidatat magna
        ipsum ut voluptate. 
        <a
          href="#"
          tabindex="-1"
        >
          E pluribus unum.
        </a>
        
      
      </p>
      
    
    </div>
    
  
  </div>
  

</body>
`;

exports[`Collapsible API start sets attributes 1`] = `
<body>
  
  
  <div
    class="collapsible"
    data-collapsible="collapsible-id"
    data-visible="true"
  >
    
    
    <h5>
      
      
      <button
        aria-controls="collapsible-id"
        aria-expanded="true"
        class="collapsible-trigger"
        data-target="collapsible-id"
        id="collapsible-trigger-id"
      >
        
        Collapsible Trigger
      
      </button>
      
    
    </h5>
    
    
    <div
      aria-hidden="false"
      aria-labelledby="collapsible-trigger-id"
      class="collapsible-content"
      id="collapsible-id"
      style="max-height: 0em;"
    >
      
      
      <p
        class="has-p"
      >
        
        Consectetur eiusmod laboris in non id tempor exercitation ipsum cupidatat magna
        ipsum ut voluptate. 
        <a
          href="#"
          tabindex="0"
        >
          E pluribus unum.
        </a>
        
      
      </p>
      
    
    </div>
    
  
  </div>
  

</body>
`;

exports[`Collapsible API stop + handleClick does not toggle collapsible on click 1`] = `
<div
  aria-hidden="false"
  aria-labelledby="collapsible-trigger-id"
  class="collapsible-content"
  id="collapsible-id"
  style="max-height: 0em;"
>
  
      
  <p
    class="has-p"
  >
    
        Consectetur eiusmod laboris in non id tempor exercitation ipsum cupidatat magna
        ipsum ut voluptate. 
    <a
      href="#"
      tabindex="0"
    >
      E pluribus unum.
    </a>
    
      
  </p>
  
    
</div>
`;
